<launch>
    <!-- 启动激光雷达节点 -->
    <node name="richbeam_lidar" pkg="lakibeam1" type="lakibeam1_scan_node" output="screen">
        <param name="frame_id" type="string" value="laser_link"/><!--frame_id设置-->
        <param name="output_topic" type="string" value="scan" /><!--topic设置-->
        <param name="inverted" type="bool" value="false"/><!--配置是否倒装,true倒装-->
        <param name="hostip" type="string" value="0.0.0.0"/><!--配置本机监听地址，0.0.0.0表示监听全部-->
        <param name="sensorip" type="string" value="192.168.8.2"/><!--配置sensor地址-->
        <param name="port" type="string" value="2368"/><!--配置本机监听端口-->
        <param name="angle_offset" type="int" value="0"/><!--配置点云旋转角度，可以是负数-->

        <param name="scanfreq" type="string" value="30" /><!--配置扫描频率，范围：10、20、25、30-->
        <param name="filter" type="string" value="3" /><!--配置滤波选项，范围：3、2、1、0 -->
        <param name="laser_enable" type="string" value="true" /><!--雷达扫描使能，范围：true、false-->
        <param name="scan_range_start" type="string" value="90" /><!--雷达扫描起始角度，范围：45~315-->
        <param name="scan_range_stop" type="string" value="270" /><!--雷达扫描结束角度，范围：45~315，结束角度必须大于起始角度-->
        <remap from="/richbeam_lidar/scan" to="/scan" />
    </node>
    <node pkg="tf" type="static_transform_publisher" name="base_link_to_laser0"
    args="0.0 0.0 0 0.0 0.0 0.0 /base_link /laser 40" />

    <!-- 启动摄像头节点 -->
    <group ns="camera">
        <node pkg="libuvc_camera" type="camera_node" name="mycam">
        <!-- Parameters used to find the camera -->
        <param name="vendor" value="0x09da"/>
        <param name="product" value="0x2700"/>
        <param name="serial" value=""/>
        <!-- If the above parameters aren't unique, choose the first match: -->
        <param name="index" value="0"/>

        <!-- Image size and type -->
        <param name="width" value="1280"/>
        <param name="height" value="720"/>
        <!-- choose whichever uncompressed format the camera supports: -->
        <param name="video_mode" value="mjpeg"/> <!-- or yuyv/nv12/mjpeg -->
        <param name="frame_rate" value="30"/>

        <param name="timestamp_method" value="start"/> <!-- start of frame -->
        <!-- <param name="camera_info_url" value="file:///tmp/cam.yaml"/> -->

        <param name="auto_exposure" value="3"/> <!-- use aperture_priority auto exposure -->
        <param name="auto_white_balance" value="true"/>
        </node>
    </group>
    <!-- OpenCV节点 -->
    <node name="open_cv_trace" 
          pkg="open_cv" 
          type="open_cv_teace" 
          output="screen"
          respawn="false">
        <!-- 可通过参数服务器设置订阅和发布的topic -->
        <param name="input_topic"  value="/camera/image_raw"/>
        <param name="output_topic" value="/image_converter/output"/>
        <param name="laser_topic" value="/scan"/>
    </node>
    <!-- 小车节点 -->
     <!-- 串口协议节点 -->
    <node name="uart_protocol_node" 
          pkg="ros_uart_protocol" 
          type="ros_uart_protocol_node" 
          output="screen"
          respawn="true"
          respawn_delay="5">


        <!-- 串口参数配置 -->
        <param name="serial_port" value="/dev/ttyACM0" />
        <param name="baudrate" value="115200" />
        
        <!-- 机器人运动参数 -->
        <param name="wheel_separation" value="0.425" />  <!-- 轮间距(米) -->
        <param name="wheel_radius" value="0.085" />      <!-- 轮半径(米) -->
    </node>

    <!-- 可选的Rviz启动配置 -->
    <param name="robot_description" textfile="$(find ros_uart_protocol)/urdf/nailong_model.urdf" />
    <!-- <node name="rviz" pkg="rviz" type="rviz" args="-d $(find ros_uart_protocol)/config/nailong.rviz"/> -->

       <!-- 启动机器人状态和关节状态发布节点 -->
    <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" />
    <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher" />

</launch>